File: vuln-sandbox/php/cwe-78/command-injection-safe.php
<?php
// CWE-78: OS Command Injection - Safe Pattern
// Rule: php-command-injection-exec
// Expected: TrueNegative

// SAFE: Using escapeshellarg() to sanitize user input
$filename = escapeshellarg($_GET['filename']);
exec("convert " . $filename . " output.png");

// SAFE: Validate input against allowlist before use
$allowed_sizes = ['800x600', '1024x768', '1920x1080'];
$size = $_POST['size'];
if (in_array($size, $allowed_sizes)) {
    exec("mogrify -resize " . escapeshellarg($size) . " image.jpg");
}

// SAFE: Static command, no user input
$output = shell_exec("ls /var/log");
echo htmlspecialchars($output);
?>

File: vuln-sandbox/ruby/cwe-78/command-injection-safe.rb
# CWE-78: OS Command Injection - Safe Pattern
# Rule: ruby-command-injection-system
# Expected: TrueNegative

class FilesController < ApplicationController
  def convert
    # SAFE: Arguments passed as array, no shell interpolation
    filename = params[:filename]
    sanitized = File.basename(filename)
    system("convert", sanitized, "output.png")
    render plain: "Converted"
  end

  def resize
    # SAFE: Using Open3 with array arguments
    require 'open3'
    size = params[:size].to_s.match?(/\A\d+x\d+\z/) ? params[:size] : "800x600"
    file = File.basename(params[:file])
    Open3.capture2("mogrify", "-resize", size, file)
  end
end
